### PR TITLE
Added decorator to fallback to plex tvdb mirror with tvdb is down

### PR DESCRIPTION
**Pull request by p0psicles**
…ptions.

- [x] PR is based on the DEVELOP branch
- [x] Don't send big changes all at once. Split up big PRs into multiple smaller PRs that are easier to manage and review
- [x] Read the [contribution guide](https://github.com/pymedusa/Medusa/blob/master/.github/CONTRIBUTING.md)

Fallback to plex mirror, works okay. I've been able to test that.
I havent tested the reverting to tvdb base_url.

I was thinking about implementing the following features:
- [ ] ~~Fallback after x minutes amount of tvdb exceptions~~
- [x] Fallback after one tvdb error
- [x] Restore tvdb after x amount of hours of fallback (default now 3 hours).
- [x] Notifications to user for when falling back and restore.
image
- [x] Make it configurable. Options for: use plex fallback, fallback after x hours, disable notifications.
- [ ] ~~Option to force move back to tvdb~~
I'm canceling this, as if I need to add it, it should be done through apiv2 rpc. And too much is changing in that area now.

### CODE REVIEW

**Comment by labrys** [comment]
Fallback should be optional (although I would default to enabled) as a user may not want to fall back to plex.

**Comment by fernandog** [comment]
Why he wouldnt want to fallback? He would prefer not have shows updated?  
I see no point creating a setting for this

File: medusa/indexers/tvdbv2/tvdbv2_api.py [hunk 1]
```diff
@@ -33,6 +34,53 @@
...
+            tvdb_connect_failures_count += 1
+            if tvdb_connect_failures_count < 4:
+                raise
+        except (MaxRetryError, RequestError) as e:
+            logger.warning("could not connect to TheTvdb.com, with reason '%s'.", e.reason)
+            tvdb_connect_failures_count += 1
+            if tvdb_connect_failures_count < 4:
+                raise
+        except Exception as e:
+            logger.warning("could not connect to TheTvdb.com, with reason '%s'.", e)
```
Comment (p0psicles) [inline medusa/indexers/tvdbv2/tvdbv2_api.py]: If you change the apikey, it will get to this, through a 401 exception. And you can (kinda) test it.

**Comment by p0psicles** [comment]
I've configured it pretty aggressive now. I'm going to add some configuration options. And then I think it's close to finished.

File: medusa/indexers/tvdbv2/tvdbv2_api.py [hunk 2]
```diff
@@ -29,10 +30,65 @@
...
+API_BASE_URL_FALLBACK = 'https://tvdb2.plex.tv'
+
+
+def plex_fallback(func):
+    """Fallback to plex if tvdb fails to connect."""
+    def inner(*args, **kwargs):
+        config = args[0].config['session'].fallback_config
+
+        def fallback_notification():
+            ui.notifications.error('Tvdb2.plex.tv fallback',
```
Comment (duramato) [inline medusa/indexers/tvdbv2/tvdbv2_api.py]: I'd also add a warning log. UI notification require the user to be on the Web UI?

Comment (fernandog) [inline medusa/indexers/tvdbv2/tvdbv2_api.py]: we already have a warning about not able to connect to tvdb. we have too much warning

**Comment by fernandog** [comment]
@p0psicles did you got this?

2017-03-14 03:00:07 ERROR    SHOWUPDATER :: [8fbc3d8] Problem running show_updater, Indexer TVDBv2 seems to be having issues while trying to get updates for show Legion. Cause: 'TVDBv2' object has no attribute 'updates_api'.
Traceback (most recent call last):
  File "/home/osmc/Medusa/medusa/show_updater.py", line 86, in run
    update_max_weeks)
  File "/home/osmc/Medusa/medusa/indexers/tvdbv2/tvdbv2_api.py", line 622, in get_last_updated_series
    updates = self.updates_api.updated_query_get(from_time).data
AttributeError: 'TVDBv2' object has no attribute 'updates_api'

**Comment by p0psicles** [comment]
Yes I noticed. Will fix that later.

**Comment by fernandog** [comment]
@p0psicles i got 5 warnings in a row:

should it really be a warning? or maybe only first one?

image

**Comment by p0psicles** [comment]
Maybe we can make it transparent and throttle it. Those are minor details.

**Comment by p0psicles** [comment]
@pymedusa/developers  Any other features that need to be in, as an MVP?

**Comment by fernandog** [comment]
im still testing. im ok with current features

are you going to wait @Thraxis mirror?

**Comment by p0psicles** [comment]
Not necessarily. But I do like to finish another Pr first. That Implements jwt token Auth hook into our current tvdb lib. Then we can remove the 'ugly' authentication code from tvdb_api. Labrys is going to fix some small issues in that branch. And then that could be merged first.

**Comment by fernandog** [comment]
@p0psicles

76 of this:

2017-03-20 03:00:09 ERROR    SHOWUPDATER :: [ccd344f] Problem running show_updater, Indexer TVDBv2 seems to be having issues while trying to get updates for show The Blacklist: Redemption. Cause: 'NoneType' object has no attribute 'request'.
Traceback (most recent call last):
  File "/home/osmc/Medusa/medusa/show_updater.py", line 86, in run
    update_max_weeks)
  File "/home/osmc/Medusa/medusa/indexers/tvdbv2/tvdbv2_api.py", line 94, in inner
    args[0]._authenticate()  # pylint: disable=W021
  File "/home/osmc/Medusa/medusa/indexers/tvdbv2/tvdbv2_api.py", line 174, in _authenticate
    access_token = auth_api.login_post(authentication_string)
  File "/home/osmc/Medusa/lib/tvdbapiv2/apis/authentication_api.py", line 124, in login_post
    callback=params.get('callback'))
  File "/home/osmc/Medusa/lib/tvdbapiv2/api_client.py", line 324, in call_api
    response_type, auth_settings, callback)
  File "/home/osmc/Medusa/lib/tvdbapiv2/api_client.py", line 150, in __call_api
    post_params=post_params, body=body)
  File "/home/osmc/Medusa/lib/tvdbapiv2/api_client.py", line 360, in request
    body=body)
  File "/home/osmc/Medusa/lib/tvdbapiv2/rest.py", line 113, in POST
    body=body)
  File "/home/osmc/Medusa/lib/tvdbapiv2/rest.py", line 63, in request
    r = self.session.request(method, url, params=query_params, headers=headers, json=body)
AttributeError: 'NoneType' object has no attribute 'request'

**Comment by p0psicles** [comment]
Yeah I recommend not to use this for a couple of days. Until we have merged the Auth changes.

File: lib/tvdbapiv2/auth/tvdb.py [hunk 3]
```diff
@@ -68,6 +68,7 @@ def login(self):
         response = requests.post(
             urljoin(self.api_base, 'login'),
             json=self.authorization,
+            verify=False,
```
Comment (p0psicles) [inline lib/tvdbapiv2/auth/tvdb.py]: Reminder remove this when ready to merge.!

File: lib/tvdbapiv2/auth/tvdb.py [hunk 4]
```diff
@@ -89,6 +90,7 @@ def refresh(self):
         response = requests.get(
             urljoin(self.api_base, 'refresh_token'),
             headers=self.auth_header,
+            verify=False,
```
Comment (p0psicles) [inline lib/tvdbapiv2/auth/tvdb.py]: Reminder remove this when ready to merge.

**Comment by p0psicles** [comment]
Major change i've made, is moving all plex fallback related code to medusa/indexers/tvdbv2/fallback.py.
Other changes are mostly cleanups and small improvements in using one Sessions instance for tvdb.

Removing the plex fallback code is not more then just removing the decorators.

**Pull request merged**